dpx/src/dpx_pdfobj.rs

```diff
@@ -1265,53 +1271,70 @@ impl pdf_dict {
             panic!("pdf_add_dict(): Passed invalid value");
         }
         /* If this key already exists, simply replace the value */
-        let mut data = self;
-        while !data.key.is_null() {
-            if key.as_ref() == (*data.key).as_name().to_bytes() {
-                /* Release the old value */
-                pdf_release_obj(data.value);
-                data.value = value;
-                return 1i32;
-            }
-            data = &mut *data.next
+        if let Some(existing) = self.inner.insert(pdf_name::new(key), value) {
+            pdf_release_obj(existing);
+            1
+        } else {
+            0
         }
-        let name = pdf_new_name(key);
-        /*
-         * We didn't find the key. We build a new "end" node and add
-         * the new key just before the end
-         */
-        let new_node = Box::new(pdf_dict::new());
-        data.next = Box::into_raw(new_node);
-        data.key = name;
-        data.value = value;
-        0i32
     }
     /* pdf_merge_dict makes a link for each item in dict2 before stealing it */
     pub unsafe fn merge(&mut self, dict2: &Self) {
```

Comment:
Does this mean the ownership of dict2 is transferred to &self? In which case we can take dict2 as Self

Reply:
I think it just increments a reference count on all of dict2’s values. Taking as Self would just mean some code has to pre-emptively clone dict2. Not advantageous.